Adding PG_CONN_STR environment variable for PG backend.

phillipsj opened the pull request:
This adds the ability to use the PG_CONN_STR environment variable to set the connection string for the postgres backed. Fixes #27655

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/terraform?pullRequest=28205)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/terraform?pullRequest=28205)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/terraform?pullRequest=28205).</sub>

remilapeyre reviewed (changes requested):
Thanks for taking the time to write a PR @phillipsj! The change looks good and I think it should get released in the next minor version once we take care of the last details :)

remilapeyre commented:
I forgot to say that the documentation should also be updated with the change.

phillipsj commented:
I rebased on the lastest master, made the updates to the variable, added the test, and updated the docs. There is a failing test which I didn't touch, so I feel that came from the latest master.

remilapeyre reviewed (approved):
Thanks 🎉 !

crw commented on internal/backend/remote-state/pg/backend_test.go:
```diff
@@ -26,8 +27,8 @@ func testACC(t *testing.T) {
-	if os.Getenv("DATABASE_URL") == "" {
-		os.Setenv("DATABASE_URL", "postgres://localhost/terraform_backend_pg_test?sslmode=disable")
+	if os.Getenv("PG_CONN_STR") == "" {
+		os.Setenv("PG_CONN_STR", "postgres://localhost/terraform_backend_pg_test?sslmode=disable")
```
Should these be `PGCONNSTR`?

crw commented on internal/backend/remote-state/pg/backend_test.go:
```diff
@@ -349,6 +350,66 @@ func TestBackendConcurrentLock(t *testing.T) {
+			if tc.asEnv {
+				os.Setenv("PG_CONN_STR", tc.pgconnstr)
+			} else {
+				os.Unsetenv("PG_CONN_STR")
```
Should these be `PGCONNSTR`?

crw commented on internal/backend/remote-state/pg/backend_test.go:
```diff
@@ -349,6 +350,66 @@ func TestBackendConcurrentLock(t *testing.T) {
+
 func getDatabaseUrl() string {
-	return os.Getenv("DATABASE_URL")
+	return os.Getenv("PG_CONN_STR")
```
Should this be `PGCONNSTR`?

crw reviewed (commented):
Did these get missed in the conversion to no-underscores, or are these underscore names acceptable for another reason?

crw commented:
@phillipsj Thanks for the contribution! Can you check out my quick review, and also resolve the merge conflicts in the docs? We did a big doc migration recently that may have caused that issue, let me know if you need help and I'll pull in the doc expert.
